src/pages/signin/SignInPage.js

```diff
@@ -51,37 +57,65 @@ class SignInPage extends Component {
         // - A login has not been entered yet
         const showLoginForm = !this.props.credentials.login;
 
-        // Show the password form if
+        // Show the old password form if
         // - A login has been entered
         // - AND an account exists and is validated for this login
         // - AND a password hasn't been entered yet
         // - AND haven't forgotten password
+        // - AND the user is NOT on the passwordless beta
         const showPasswordForm = this.props.credentials.login
             && this.props.account.validated
             && !this.props.credentials.password
-            && !this.props.account.forgotPassword;
+            && !this.props.account.forgotPassword
+            && !Permissions.canUsePasswordlessLogins(this.props.betas);
+
+        // Show the new magic code / validate code form if
+        // - A login has been entered
+        // - AND the user is on the 'passwordless' beta
+        const showValidateCodeForm = this.props.credentials.login
+            && Permissions.canUsePasswordlessLogins(this.props.betas);
 
         // Show the resend validation link form if
         // - A login has been entered
         // - AND is not validated or password is forgotten
-        const shouldShowResendValidationLinkForm = this.props.credentials.login
-            && (!this.props.account.validated || this.props.account.forgotPassword);
+        // - AND user is not on 'passwordless' beta
+        const showSendValidateCodeForm = this.props.credentials.login
+            && (!this.props.account.validated || this.props.account.forgotPassword)
+            && !Permissions.canUsePasswordlessLogins(this.props.betas);
 
-        const welcomeText = shouldShowResendValidationLinkForm
-            ? ''
-            : this.props.translate(`welcomeText.${showPasswordForm ? 'welcomeBack' : 'welcome'}`);
+        let welcomeText;
+        if (showValidateCodeForm) {
+            if (this.props.account.requiresTwoFactorAuth) {
+                // We will only know this after a user signs in successfully, without their 2FA code
+                welcomeText = this.props.translate('validateCodeForm.enterTwoFactorOrRecoveryCode');
+            } else if (this.props.account.validated) {
+                welcomeText = this.props.translate('welcomeText.welcomeBackEnterMagicCode', {login: this.props.credentials.login});
+            } else if (!this.props.account.validated) {
+                welcomeText = this.props.translate('welcomeText.welcomeEnterMagicCode', {login: this.props.credentials.login});
+            }
+        } else if (showSendValidateCodeForm) {
+            welcomeText = '';
+        } else if (showPasswordForm) {
+            welcomeText = this.props.translate('welcomeText.welcomeBack');
+        } else {
+            welcomeText = this.props.translate('welcomeText.welcome');
+        }
 
```

Comment:
NAB: I think you can do this instead and it's slightly shorter (also some stylistic preference here though) 

```JS
        let welcomeText = '';
        if (showValidateCodeForm) {
            if (this.props.account.requiresTwoFactorAuth) {
                // We will only know this after a user signs in successfully, without their 2FA code
                welcomeText = this.props.translate('validateCodeForm.enterTwoFactorOrRecoveryCode');
            } else {
                welcomeText = this.props.account.validated
                    ? this.props.translate('welcomeText.welcomeBackEnterMagicCode', {login: this.props.credentials.login})
                    : this.props.translate('welcomeText.welcomeEnterMagicCode', {login: this.props.credentials.login});
            }
        } else if (showPasswordForm) {
            welcomeText = this.props.translate('welcomeText.welcomeBack');
        } else {
            welcomeText = this.props.translate('welcomeText.welcome');
        }
```

Reply:
I like this cleanup - though I think the last `else` needs to be `else if (!showResendValidationForm) {` to make sure all the cases still line up